File: network.py
import math
from PySide import QtCore

import node
import tdvector

class Network():

    def __init__(self, timeStep = 4.0):

            self.nodes = []
            self.connections = []
            self.timeStep = timeStep
            self.repelFactor = 10.0 ** 2.0
            self.attractFactor = 10.0 ** -4.0
            self.dampFactor = 0.5

    def addNode(self, xPos, yPos):
        self.nodes.append(node.Node(xPos, yPos))

    def connect(self, node1, node2):
        self.connections.append([node1, node2])
        node1.connect(node2)

    def calculateForce(self):

        i = 0
        for inode in self.nodes[:-1]:
            for jnode in self.nodes[i+1:]:

                deltaPos = tdvector.difference(jnode.position, inode.position)
                theta = deltaPos.angle()
                forceMag = (self.repelFactor / deltaPos.magnitudeSquare())
                forceVec = tdvector.TDVector(forceMag * math.cos(theta),
                                             forceMag * math.sin(theta))
                inode.forceSum.subtract(forceVec)
                jnode.forceSum.add(forceVec)
            i += 1

        for connection in self.connections:
            deltaPos = tdvector.difference(connection[1].position, connection[0].position)
            theta = deltaPos.angle()
            forceMag = deltaPos.magnitude() * self.attractFactor
            forceVec = tdvector.TDVector(forceMag * math.cos(theta),
                                         forceMag * math.sin(theta))

            connection[0].forceSum.add(forceVec)
            connection[1].forceSum.subtract(forceVec)


    def move(self):
        for node in self.nodes:
            node.velocity.add(node.forceSum.scale(self.timeStep))
            node.dampVelocity(self.dampFactor)
            node.position.add(node.velocity.scale(self.timeStep))
            node.zero()

File: Network.cpp
#include <algorithm>
#include <cmath>
#include "Network.h"

using namespace Linear;
using namespace std;

Network::Network() {
}

Network::~Network() {
}

void Network::calculate_forces(float time_step, float repel_factor,
                               float attract_factor, float damp_factor) {
    for (int i = 0; i < num_nodes(); ++i) {
        // Repulsive force
        for (int j = i + 1; j < num_nodes(); ++j) {
            Vec2f delta_pos = nodes[j]->pos - nodes[i]->pos;
            float theta = delta_pos.theta();
            float force_mag = repel_factor / delta_pos.magnitude_square();
            Vec2f force_vec(theta);
            force_vec *= force_mag;
            nodes[i]->force -= force_vec;
            nodes[j]->force += force_vec;
        }

        // Dampening
        Vec2f force_vec = nodes[i]->vel * damp_factor;
        nodes[i]->force -= force_vec;
    }

    // Elastic force
    for (auto& connection: connections) {
        Node* n1 = connection.node_a;
        Node* n2 = connection.node_b;
        Vec2f delta_pos = n1->pos - n2->pos;
        float theta = delta_pos.theta();
        float force_mag = delta_pos.magnitude() * attract_factor;
        Vec2f force_vec(theta);
        force_vec *= force_mag;
        n2->force += force_vec;
        n1->force -= force_vec;
    }
}

void Network::move_nodes(float time_step) {
    for (Node*& node: nodes) {
        node->vel += node->force * time_step;
        node->pos += node->vel * time_step;
        node->force.zero();
    }
}

void Network::add_node(Node* node) {
    nodes.push_back(node);
}

void Network::add_node(float x_pos, float y_pos) {
    nodes.push_back(new Node(x_pos, y_pos));
}

int Network::num_nodes() {
    return nodes.size();
}

void Network::step(float time_step, float repel_factor,
                   float damp_factor, float attract_factor) {
    calculate_forces(time_step, repel_factor, attract_factor, damp_factor);
    move_nodes(time_step);
}

void Network::connect(Node * node_a, Node* node_b) {
    node_a->connect(node_b);
    connections.push_back({node_a, node_b});
}

